pkg/proxy/iptables/proxier_test.go

```diff
@@ -1345,6 +1347,423 @@ func assertIPTablesRulesNotEqual(t *testing.T, expected, result string) {
+		}
+		isProbabilisticMatch := (match[1] != "")
+		target := match[2]
+		natDestination := match[4]
+
+		if !tracer.ruleMatches(rule, sourceIP, destIP, destPort) {
+			continue
+		}
+		tracer.matches = append(tracer.matches, rule)
+
+		switch target {
+		case "KUBE-MARK-MASQ":
+			tracer.markMasq = true
+			continue
+
+		case "KUBE-MARK-DROP":
+			tracer.markDrop = true
+			continue
+
+		case "ACCEPT", "REJECT":
+			tracer.outputs = append(tracer.outputs, target)
+			return true
+
+		case "DNAT":
+			tracer.outputs = append(tracer.outputs, natDestination)
+			return true
+
+		default:
+			// We got a "-j KUBE-SOMETHING", so process that chain
+			terminated := tracer.runChain(table, target, sourceIP, destIP, destPort)
+
+			// If the subchain hit a terminal rule AND the rule that sent us
+			// to that chain was non-probabilistic, then this chain terminates
+			// as well. But if we went there because of a --probability rule,
+			// then we want to keep accumulating further matches against this
+			// chain.
+			if terminated && !isProbabilisticMatch {
+				return true
+			}
+		}
+	}
+
+	return false
+}
+
+// tracePacket determines what would happen to a packet with the given sourceIP, destIP,
+// and destPort, given the indicated iptables ruleData. nodeIP is the local node IP (for
+// rules matching "LOCAL").
+//
+// The return values are: an array of matched rules (for debugging), the final packet
+// destinations (a comma-separated list of IPs, or one of the special targets "ACCEPT",
+// "DROP", or "REJECT"), and whether the packet would be masqueraded.
+func tracePacket(ruleData, sourceIP, destIP, destPort, nodeIP string) ([]string, string, bool) {
+	tracer, err := newIPTablesTracer(ruleData, nodeIP)
+	if err != nil {
+		panic("bad iptables ruleData: " + err.Error())
+	}
+
+	// nat:PREROUTING goes first, then the filter chains, then nat:POSTROUTING. For
+	// our purposes that means we run through the "nat" chains first, starting at
+	// KUBE-SERVICES, then we do the "filter" chains. The only interesting thing that
+	// happens in nat:POSTROUTING is that the masquerade mark gets turned into actual
+	// masquerading.
+
+	tracer.runChain("nat", "KUBE-SERVICES", sourceIP, destIP, destPort)
+
+	// Process pending DNAT (which theoretically might affect REJECT/ACCEPT filter rules)
+	if len(tracer.outputs) != 0 {
+		destIP = tracer.outputs[0]
+	}
+
+	// Now run the filter rules to see if the packet is REJECTed or ACCEPTed. The DROP
+	// rule is created by kubelet, not us, so we have to simulate that manually
+	if tracer.markDrop {
+		return tracer.matches, "DROP", false
+	}
+	tracer.runChain("filter", "KUBE-SERVICES", sourceIP, destIP, destPort)
+	tracer.runChain("filter", "KUBE-EXTERNAL-SERVICES", sourceIP, destIP, destPort)
+	tracer.runChain("filter", "KUBE-FORWARD", sourceIP, destIP, destPort)
+	tracer.runChain("filter", "KUBE-NODEPORTS", sourceIP, destIP, destPort)
+
+	return tracer.matches, strings.Join(tracer.outputs, ", "), tracer.markMasq
+}
```

Comment:
The above was surprisingly coherent - well done.  Now that I reach the bottom, I feel like we need a suite of tests to prove that tracePacket() works against known inputs - is that not needed for some reason I am missing?

Reply:
added